miner/proposal_builder.go

```diff
@@ -374,137 +335,370 @@ func (pb *ProposalBuilder) decideMeshHash(ctx context.Context, current types.Lay
 	return mesh
 }
 
-func (pb *ProposalBuilder) handleLayer(ctx context.Context, layerID types.LayerID) error {
-	var (
-		beacon types.Beacon
-		err    error
-		epoch  = layerID.GetEpoch()
-	)
-
-	if layerID <= types.GetEffectiveGenesis() {
-		return errGenesis
+func (pb *ProposalBuilder) initSessionData(ctx context.Context, lid types.LayerID) error {
+	if pb.session == nil || pb.session.epoch != lid.GetEpoch() {
+		pb.session = &session{epoch: lid.GetEpoch()}
 	}
-	if !pb.syncer.IsSynced(ctx) {
-		return errNotSynced
+	if pb.session.atx == types.EmptyATXID {
+		atx, err := atxs.GetByEpochAndNodeID(pb.cdb, pb.session.epoch-1, pb.signer.NodeID())
+		if err != nil {
+			if errors.Is(err, sql.ErrNotFound) {
+				err = errAtxNotAvailable
+			}
+			return fmt.Errorf("get atx in epoch %v: %w", pb.session.epoch-1, err)
+		}
+		pb.session.atx = atx.ID()
+		pb.session.atxWeight = atx.GetWeight()
 	}
-
-	// make sure the miner is eligible first
-	nonce, err := pb.nonceFetcher.VRFNonce(pb.signer.NodeID(), layerID.GetEpoch())
-	if err != nil {
+	if pb.session.nonce == 0 {
+		nonce, err := pb.cdb.VRFNonce(pb.signer.NodeID(), pb.session.epoch)
+		if err != nil {
+			return fmt.Errorf("missing nonce: %w", err)
+		}
+		pb.session.nonce = nonce
+	}
+	if pb.session.beacon == types.EmptyBeacon {
+		beacon, err := beacons.Get(pb.cdb, pb.session.epoch)
+		if err != nil || beacon == types.EmptyBeacon {
+			return fmt.Errorf("missing beacon for epoch %d", pb.session.epoch)
+		}
+		pb.session.beacon = beacon
+	}
+	if pb.session.prev == 0 {
+		prev, err := ballots.LastInEpoch(pb.cdb, pb.session.atx, pb.session.epoch)
+		if err != nil && !errors.Is(err, sql.ErrNotFound) {
+			return err
+		}
+		if err == nil {
+			pb.session.prev = prev.Layer
+		}
```

Comment:
NIT: This is more readable IMO:
```suggestion
		switch {
		case err == nil:
			pb.session.prev = prev.Layer
		case errors.Is(err, sql.ErrNotFound):
		default:
			return err
		}
```